### PR TITLE
Schedules

### LINKED ISSUE #810: Integrate Scheduling workflows in Task List
As a developer, I'd like to integrate the following scheduling workflows into Task list page. 

 - [x] Global actions
 - [x] Group actions
 - [x] Table in-line actions
 - [x] Execution status per row-item
 - [x] Consistent UI/UX with rest of the list pages
 - [ ] Test coverage

child of #809

### LINKED ISSUE #811: Add Schedule List
As a developer, I'd like to add a list page for Schedules, so we can navigate to, inspect, and monitor the available schedules. 

 - [x] Global actions
 - [x] Group actions
 - [x] Table in-line actions
 - [x] Execution status per row-item
~- [x] Consistent UI/UX with rest of the list pages~
 - [ ] Test coverage

child of #809

### LINKED ISSUE #812: Add Schedule Details
As a developer, I'd like to add a details page for a particular Schedule, so we can navigate to, inspect, and monitor the scheduled parameters such as cron-expression.

 - [x] Schedule details
 - [x] Execution status per schedule
 - [x] Tabular model for consistent UX with rest of the Detail pages
 - [ ] Plaecholder for metrics/status per schedule
 - [x] Navigation mechanism between Task List and each Schedules and vice-versa
 - [ ] Test coverage

child of #809

### LINKED ISSUE #813: Add support for multiple-schedules per each Task
As a developer, I'd like to add support to schedule the same Task "n" no. of times with different criteria. 

 - [x] List available schedules per Task
 - [x] UX for single vs. multiple schedules
 - [x] Navigation mechanism between Task List and each Schedules and vice-versa
 - [ ] Test coverage

child of #809

### PR DESCRIPTION
**Pull request by oodamien**
+ Add Schedule List
+ Add Schedule Details
+ Add support for multiple-schedules per each Task
+ Integrate Scheduling workflows in Task List

Resolves #810 
Resolves #811 
Resolves #812 
Resolves #813

To Talk:
- Counter (slow): remove?

### CODE REVIEW

**Comment by ghillert** [comment]
Looks like error message may not show up up correctly: 
image

File: ui/src/app/tasks/task-schedule-create/task-schedule-create.component.html [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+              </td>
+            </tr>
+            </tbody>
+          </table>
+        </div>
+      </div>
+    </div>
+    <div class="row form-group" [class.has-error]="form.get('cron').invalid">
+      <div class="col-xs-4">
+        <label class="control-label" for="cron">Cron Expression:</label>
```
Comment (ghillert) [inline ui/src/app/tasks/task-schedule-create/task-schedule-create.component.html]: We should provide some usage examples for users. Who remembers the CRON syntax? ;-)

File: ui/src/app/tasks/task-schedule-create/task-schedule-create.validator.spec.ts [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+    it('valid', () => {
+      [
+        null,
+        '0 0 0 0 0 0 0',
+        '0 15 10 * * ? 2005',
+        '0 15 10 ? * 6L 2002-2005',
+        '0 15 10 ? * MON-FRI',
+        '0 15 10 ? * 6#3',
+        '0 11 11 11 11 ?',
+        '0 0 12 * * ?',
```
Comment (ghillert) [inline ui/src/app/tasks/task-schedule-create/task-schedule-create.validator.spec.ts]: According the PCF docs, cron expressions with more than 5 fields should be disallowed. Currently it even throws a nasty exception. See: https://docs.pivotal.io/pcf-scheduler/1-2/using-jobs.html#create-jobs under `Schedule a Job`

File: ui/src/app/tasks/task-schedule-create/task-schedule-create.component.html (hunk 1 above)
Comment (ghillert) [reply ui/src/app/tasks/task-schedule-create/task-schedule-create.component.html]: Also maybe detail the syntax according to PCF docs.

File: ui/src/app/tasks/task-schedule-create/task-schedule-create.component.html [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+          <label>Task Name:</label>
+        </div>
+        <div class="col-xs-20">
+          <a style="cursor: pointer" (click)="detailsTask(schedule.taskDefinitions[0])">
+            {{ schedule.taskDefinitions[0] }}</a>
+        </div>
+      </div>
+      <div class="row form-group" [class.has-error]="form.get('names').controls[0].invalid">
+        <div class="col-xs-4">
+          <label class="control-label" for="name">Scheduler name:</label>
```
Comment (cppwfs) [inline ui/src/app/tasks/task-schedule-create/task-schedule-create.component.html]: Probably want to rename this to Schedule name:

File: ui/src/app/tasks/task-schedule-create/task-schedule-create.validator.spec.ts (hunk 2 above)
Comment (sabbyanandan) [reply ui/src/app/tasks/task-schedule-create/task-schedule-create.validator.spec.ts]: @ghillert: We aren't going to be validating anything on the client side. We can, however, point to an external link (in cf docs) for someone wanting to learn more about the syntax correctness.

**Comment by ghillert** [comment]
Requesting a non-existing schedule results in a stale state instead of providing an error response notification.

image

File: ui/src/app/tasks/task-schedules-destroy/task-schedules-destroy.component.html [hunk 4]
```diff
@@ -0,0 +1,48 @@
+<div *ngIf="taskSchedules">
+  <div class="modal-header">
+    <h4 class="modal-title pull-left">Confirm Delete Task Schedules</h4>
```
Comment (ghillert) [inline ui/src/app/tasks/task-schedules-destroy/task-schedules-destroy.component.html]: Should rename. Not exactly sure to what - Maybe (too wordy??): "Confirm the Deletion of Task Schedules"

image

File: ui/src/app/tasks/task-schedules-destroy/task-schedules-destroy.component.html [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+          <div *ngIf="item.cronExpression">{{ item.cronExpression }}</div>
+        </td>
+      </tr>
+      </tbody>
+    </table>
+  </div>
+  <div class="modal-footer" style="text-align:center">
+    <button id="btn-cancel" type="button" class="btn btn-default" (click)="cancel()">No</button>
+    <button id="btn-destroy" type="button" class="btn btn-primary" (click)="destroy(taskSchedules)">
+      Delete Task Schedule(s)
```
Comment (ghillert) [inline ui/src/app/tasks/task-schedules-destroy/task-schedules-destroy.component.html]: Maybe test for singular and plural?

File: ui/src/app/tasks/task-schedules-destroy/task-schedules-destroy.component.ts [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+
+  /**
+   * Submit destroy task schedule(s)
+   */
+  destroy() {
+    this.loggerService.log(`Proceeding to delete ${this.taskSchedules.length} task schedule(s).`, this.taskSchedules);
+    const busy = this.tasksService.destroySchedules(this.taskSchedules)
+      .pipe(takeUntil(this.ngUnsubscribe$))
+      .subscribe((data) => {
+        this.notificationService.success(`${data.length} task schedule(s) delete.`);
```
Comment (ghillert) [inline ui/src/app/tasks/task-schedules-destroy/task-schedules-destroy.component.ts]: Rename `delete` to `deleted`. Maybe test for singular/plural in case only 1 task schedule is being deleted?

**Comment by ghillert** [comment]
@oodamien @sabbyanandan Feels like a more serious UX issue:

On the Task Definition List page, there is no indication that a specific task definition is associated with a schedule. We should at least disable the task-schedule-delete button in case no schedules exist for the task definition.

image

File: ui/src/app/tasks/task-definitions/task-definitions.component.html [hunk 7]
```diff
@@ -111,6 +107,15 @@ <h1 class="no-user-selection">Tasks</h1>
                       class="btn btn-default" style="margin-left: 0;" title="Launch">
                 <span class="glyphicon glyphicon-play"></span>
               </button>
+              <button *ngIf="schedulerEnabled" type="button" name="task-schedule{{ i }}" (click)="schedule(item)" [appRoles]="['ROLE_CREATE']"
+                      class="btn btn-default" style="margin-left: 0;" title="Schedule">
+                <span class="glyphicon glyphicon-time"></span>
+              </button>
+              <button *ngIf="schedulerEnabled" type="button" name="task-schedule-destroy{{ i }}" (click)="destroySchedules(item)"
+                      [appRoles]="['ROLE_CREATE']" class="btn btn-ico-remove-schedules btn-default" style="margin-left: 0;" title="Schedule">
```
Comment (ghillert) [inline ui/src/app/tasks/task-definitions/task-definitions.component.html]: Rename to "Remove Schedule" or "Destroy Schedule"

File: ui/src/app/tasks/task-definitions/task-definitions.component.html [hunk 8]
```diff
@@ -111,6 +107,15 @@ <h1 class="no-user-selection">Tasks</h1>
...
+                <span class="glyphicon glyphicon-time"></span>
+                <span class="glyphicon glyphicon-remove"></span>
+              </button>
               <button name="task-destroy{{ i }}" type="button" [appRoles]="['ROLE_CREATE']" (click)="destroy(item)"
                       class="btn btn-default" style="margin-left: 0;" title="Destroy">
```
Comment (ghillert) [inline ui/src/app/tasks/task-definitions/task-definitions.component.html]: Since we now have a button to remove Task Schedules, we should clarify this and maybe rename to "Destroy Task Definition"

Comment (ghillert) [inline ui/src/app/tasks/task-definitions/task-definitions.component.html]: Since we now have a button to remove Task Schedules, we should clarify this and maybe rename to "Destroy Task Definition"

File: ui/src/app/tasks/task-definition/schedules/task-definition-schedule.component.html [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+    <table *ngIf="schedules.page?.items && schedules.page.items.length > 0"
+           class="table table-checkbox table-hover" id="taskSchedulesTable">
+      <thead>
+      <tr>
+        <th style="width: 30px">
+          <app-master-checkbox (change)="changeCheckboxes(schedules.page)" *ngIf="form?.checkboxes"
+                               [appRoles]="['ROLE_CREATE']" [items]="form.checkboxes"></app-master-checkbox>
+        </th>
+        <th>
+          Scheduler Name
```
Comment (ghillert) [inline ui/src/app/tasks/task-definition/schedules/task-definition-schedule.component.html]: Rename to `Schedule Name`

File: ui/src/app/tasks/task-schedule-create/task-schedule-create.component.html (hunk 3 above)
Comment (ghillert) [reply ui/src/app/tasks/task-schedule-create/task-schedule-create.component.html]: +1

File: ui/src/app/tasks/task-schedule/summary/task-schedule-summary.component.html [hunk 10]
```diff
@@ -0,0 +1,43 @@
+<div *ngIf="schedule$ | async as schedule; else loading">
+  <div>
+    <div class="row schedule-summary-row">
+      <div class="col-md-3">
+        <strong>Scheduler name:</strong>
```
Comment (ghillert) [inline ui/src/app/tasks/task-schedule/summary/task-schedule-summary.component.html]: Rename to `Schedule name`

File: ui/src/app/tasks/task-schedules/task-schedules.component.html [hunk 11]
```diff
@@ -0,0 +1,137 @@
...
+    <table *ngIf="taskSchedules && (taskSchedules.items | filterSchedules: q).length > 0"
+           class="table table-hover table-checkbox" id="taskSchedulesTable">
+      <thead>
+      <tr>
+        <th style="width: 30px">
+          <app-master-checkbox (change)="changeCheckboxes()" *ngIf="form?.checkboxes" [appRoles]="['ROLE_CREATE']"
+                               [items]="form.checkboxes"></app-master-checkbox>
+        </th>
+        <th>
+          Scheduler Name
```
Comment (ghillert) [inline ui/src/app/tasks/task-schedules/task-schedules.component.html]: Rename to `Schedule Name`

File: ui/src/app/tasks/task-schedules/task-schedules.filter.ts [hunk 12]
```diff
@@ -0,0 +1,32 @@
+import { Pipe, PipeTransform } from '@angular/core';
+import { TaskSchedule } from '../model/task-schedule';
+
+/**
+ * Filters the schedules by matching the scheduler name or the related task definition name to the input
```
Comment (ghillert) [inline ui/src/app/tasks/task-schedules/task-schedules.filter.ts]: rename `scheduler name` to `schedule name`

File: ui/src/app/tasks/task-definitions/task-definitions.component.ts [hunk 13]
```diff
@@ -286,6 +323,51 @@ export class TaskDefinitionsComponent implements OnInit, OnDestroy {
...
+    this.tasksService
+      .getSchedules({
+        task: taskDefinition.name,
+        sort: 'SCHEDULE_ID',
+        order: OrderParams.ASC,
+        page: 0,
+        size: 1000
+      }).pipe(map((page: Page<TaskSchedule>) => {
+      if (page.totalElements === 0) {
+        this.notificationService.error('No scheduler for this task.');
```
Comment (ghillert) [inline ui/src/app/tasks/task-definitions/task-definitions.component.ts]: Rename `No scheduler for this task.` to `No schedule exists for this task.`

File: ui/src/app/tasks/task-schedules-destroy/task-schedules-destroy.component.html (hunk 4 above)
Comment (oodamien) [reply ui/src/app/tasks/task-schedules-destroy/task-schedules-destroy.component.html]: @ghillert We chose 'Delete a schedule'

File: ui/src/app/tasks/task-schedules-destroy/task-schedules-destroy.component.html (hunk 5 above)
Comment (oodamien) [reply ui/src/app/tasks/task-schedules-destroy/task-schedules-destroy.component.html]: same practice than the others Delete actions

File: ui/src/app/tasks/task-schedules/task-schedules.component.html [hunk 14]
```diff
@@ -0,0 +1,137 @@
...
+          <td>
+            <a class="link-task" (click)="taskDetails(item)" style="cursor: pointer">{{ item.taskName }}</a>
+          </td>
+          <td>
+            {{ item.cronExpression }}
+          </td>
+          <td class="table-actions" width="50px" nowrap="">
+            <div class="actions" role="group">
+              <button type="button" name="schedule-destroy{{ i }}" (click)="destroySchedules([item])"
+                      class="btn btn-default" title="Details">
```
Comment (ghillert) [inline ui/src/app/tasks/task-schedules/task-schedules.component.html]: Change to delete - will do on merge

**Comment by ghillert** [comment]
Some minor polishing - merged with commit e338366286d3458f808700c4047de0fbc6101038
